### PR TITLE
fix bug in deep fp8 zero

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Fix 2 bugs in fp8 zero optimizer.
MS-AMP will crash when:
- there are param groups which don't have high precision parameter
-  partition 0 has no parameters.

### CODE REVIEW

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py [hunk 1]
```diff
@@ -139,7 +155,11 @@ def _pad_and_flat(self, values_partitions, group_fp8_mems, group_id):
             torch.Tensor: flat fp8 groups.
         """
         partition_size = dist.get_world_size(group=self.dp_process_group)
-        ref_value = values_partitions[0][0]
+        ref_value = None
+        for partition in values_partitions:
+            if len(partition) > 0:
+                ref_value = partition[0]
+                break
         dtype = ref_value.dtype
```
[the remark this answers is not in the document]
Comment (wkcn) [reply msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py:163]: We can skip the assertion statement by the following code:
```python
if ref_value is not None:
    assert all(v.dtype == dtype for v in chain(*values_partitions))
```

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py [hunk 2]
```diff
@@ -777,7 +797,12 @@ def all_gather_fp8_metas(self):
                 continue
             partition_size = len(params_partitions)
             scale_invs_partitions = [[p.meta.scale_inv for p in ps] for ps in params_partitions]
-            ref_scale = scale_invs_partitions[0][0]
+            ref_scale = None
+            for partition in scale_invs_partitions:
+                if len(partition) > 0:
+                    ref_scale = partition[0]
+                    break
+
```
Comment (wkcn) [inline msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py:805]: Please change the code to skip when ref_scale is None.
```python
if ref_scale is not None:
  scale_invs_partitions[pi].append(ref_scale.new_empty((pad, )))
```

**Review by wkcn — COMMENTED** [review]
The cases when `ref_value` is None and `ref_scale` is None should be handled.

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tocean) [reply msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py]: Good suggestion. Fixed.

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py (hunk 2 above)
Comment (tocean) [reply msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py]: Sure, removed `ref_value`.

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**